pkg/artifacts/structured.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2023 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package artifacts
+
+import (
+	"context"
+	"strings"
+
+	"github.com/tektoncd/chains/pkg/chains/objects"
+	"knative.dev/pkg/logging"
+)
+
+// StructuredSignable contains info for signable targets to become either subjects or materials in
+// intoto Statements. They are also used in SBOM signing.
+// URI is the resource uri for the target needed iff the target is a material.
+// Digest is the target's SHA digest.
+// SBOMURI is the resource uri for the target's SBOM .
+// SBOMFormat is the format of the target's SBOM.
+type StructuredSignable struct {
+	URI        string
+	Digest     string
+	SBOMURI    string
+	SBOMFormat string
+}
+
+type structuredSignableExtractor struct {
+	uriSuffix        string
+	digestSuffix     string
+	sbomURISuffix    string
+	sbomFormatSuffix string
+	isValid          func(StructuredSignable) bool
+}
+
+func (b *structuredSignableExtractor) extract(ctx context.Context, obj objects.TektonObject) []StructuredSignable {
+	logger := logging.FromContext(ctx)
```

Comment:
Extracted the StructuredSignable piece to https://github.com/tektoncd/chains/pull/924. Hopefully, this will make it easier to also do this for the object param/results.